farming/logics/traits/master_chef/farming.rs

```diff
@@ -0,0 +1,600 @@
+                &mut self.data::<Data>().arsw_token,
+                to,
+                pending_arsw,
+                Vec::new(),
+            )?;
+        }
+
+        if let Some(rewarder_address) = self.get_rewarder(pool_id) {
+            RewarderRef::on_arsw_reward(
+                &rewarder_address,
+                pool_id,
+                caller,
+                to,
+                pending_arsw,
+                user.amount,
+            )?;
+        }
+
+        self._emit_harvest_event(caller, pool_id, pending_arsw);
+        Ok(())
+    }
+
+    #[ink(message)]
+    fn withdraw_and_harvest(
+        &mut self,
+        pool_id: u32,
+        amount: Balance,
+        to: AccountId,
+    ) -> Result<(), FarmingError> {
+        let pool = self
+            .get_pool_infos(pool_id)
+            .ok_or(FarmingError::PoolNotFound)?;
+        self._update_pool(pool_id)?;
+        let caller = Self::env().caller();
+        let user = self
+            .get_user_info(pool_id, caller)
+            .ok_or(FarmingError::UserNotFound)?;
+
+        let accumulated_arsw = user
+            .amount
+            .checked_mul(pool.acc_arsw_per_share)
+            .ok_or(FarmingError::MulOverflow13)?
+            / ACC_ARSW_PRECISION;
+
+        let pending_arsw = accumulated_arsw
+            .checked_add_signed(-user.reward_debt)
+            .ok_or(FarmingError::AddOverflow15)?;
+
+        let user_reward_debt = accumulated_arsw
+            .checked_sub(
+                (casted_mul(amount, pool.acc_arsw_per_share) / ACC_ARSW_PRECISION)
+                    .try_into()
+                    .map_err(|_| FarmingError::CastTou128Error7)?,
+            )
+            .ok_or(FarmingError::SubUnderflow10)?;
+
+        let user_amount = user
+            .amount
+            .checked_sub(amount)
+            .ok_or(FarmingError::AddOverflow16)?;
+
+        self.data::<Data>().user_info.insert(
+            &(pool_id, to),
+            &UserInfo {
+                reward_debt: user_reward_debt
+                    .try_into()
+                    .map_err(|_| FarmingError::CastToi128Error7)?,
+                amount: user_amount,
+            },
+        );
+
+        PSP22Ref::transfer(
+            &mut self.data::<Data>().arsw_token,
+            to,
+            pending_arsw,
+            Vec::new(),
+        )?;
+        if let Some(rewarder_address) = self.get_rewarder(pool_id) {
+            RewarderRef::on_arsw_reward(
+                &rewarder_address,
+                pool_id,
+                caller,
+                to,
+                pending_arsw,
+                user.amount,
+            )?;
+        }
+        let lp_token = self
+            .get_lp_token(pool_id)
+            .ok_or(FarmingError::LpTokenNotFound)?;
+        PSP22Ref::transfer(&lp_token, to, amount, Vec::new())?;
+
+        self._emit_harvest_event(caller, pool_id, pending_arsw);
+        self._emit_withdraw_event(caller, pool_id, amount, to);
```

Comment:
```suggestion
        let accumulated_arsw =
            casted_mul(user.amount, pool.acc_arsw_per_share) / ACC_ARSW_PRECISION;

        let pending_arsw = <U256 as TryInto<u128>>::try_into(accumulated_arsw)
            .map_err(|_| FarmingError::CastTou128Error)?
            .checked_add_signed(-user.reward_debt)
            .ok_or(FarmingError::AddOverflow15)?;

        let user_reward_debt: i128 = accumulated_arsw
            .checked_sub(casted_mul(amount, pool.acc_arsw_per_share) / ACC_ARSW_PRECISION)
            .ok_or(FarmingError::SubUnderflow10)?
            .try_into()
            .map_err(|_| FarmingError::CastTou128Error)?;

        let user_amount = user
            .amount
            .checked_sub(amount)
            .ok_or(FarmingError::SubUnderflow)?;

        self.data::<Data>().user_info.insert(
            &(pool_id, to),
            &UserInfo {
                reward_debt: user_reward_debt,
                amount: user_amount,
            },
        );

        PSP22Ref::transfer(
            &self.data::<Data>().arsw_token,
            to,
            pending_arsw,
            Vec::new(),
        )?;
        if let Some(rewarder_address) = self.get_rewarder(pool_id) {
            RewarderRef::on_arsw_reward(
                &rewarder_address,
                pool_id,
                caller,
                to,
                pending_arsw,
                user_amount,
            )?;
        }
        let lp_token = self
            .get_lp_token(pool_id)
            .ok_or(FarmingError::LpTokenNotFound)?;
        PSP22Ref::transfer(&lp_token, to, amount, Vec::new())?;

        self._emit_withdraw_event(caller, pool_id, amount, to);
        self._emit_harvest_event(caller, pool_id, pending_arsw);
```
- Use casted_mul for accumulated_arsw as like others
- user.amount -> user_amount for rewarder argument
- For user_amount calculation, the error for checked_sub should be subtraction underflow, not about addition overflow
- Change order of event emissions (logically, withdraw action is followed by harvest)
- For `PSP22Ref::transfer` token account id doesn't need to be mutable